File: ADT/DSU/990.py
class Solution:
    def __init__(self):
        self.parent = []
        self.rank = []

    def find(self, i: int) -> int:
        if self.parent[i] != i:
            self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def Union(self, x: int, y: int) -> None:
        p_x = self.find(x)
        p_y = self.find(y)

        if p_x != p_y:
            if self.rank[p_x] > self.rank[p_y]:
                self.parent[p_y] = p_x
            elif self.rank[p_x] < self.rank[p_y]:
                self.parent[p_x] = p_y
            else:
                self.parent[p_x] = p_y
                self.rank[p_y] += 1

    def equationsPossible(self, equations: List[str]) -> bool:
        self.parent = list(range(26))
        self.rank = [1] * 26

        for s in equations:
            if s[1] == "=":
                self.Union(ord(s[0]) - ord("a"), ord(s[3]) - ord("a"))

        for s in equations:
            if s[1] == "!":
                if self.find(ord(s[0]) - ord("a")) == self.find(ord(s[3]) - ord("a")):
                    return False
        return True

File: ADT/DSU/990.js
// Tc = O(E α(V)) =~ O(V+E)
// Where α(V) is the inverse Ackermann function → grows extremely slowly
// (for all practical purposes ≤ 4).


/**
 * @param {string[]} equations
 * @return {boolean}
 */
var equationsPossible = function(equations) {
    const parent = new Array(26).fill(0).map((_, i) => i);
    const rank = new Array(26).fill(1);

    const find = (x) => {
        if (parent[x] !== x) {
            parent[x] = find(parent[x]);  // path compression
        }
        return parent[x];
    };

    const union = (x, y) => {
        const px = find(x);
        const py = find(y);
        if (px !== py) {
            if (rank[px] > rank[py]) {
                parent[py] = px;
            } else if (rank[px] < rank[py]) {
                parent[px] = py;
            } else {
                parent[px] = py;
                rank[py] += 1;
            }
        }
    };

    // First process all "==" equations
    for (const eq of equations) {
        if (eq[1] === "=") {
            const a = eq.charCodeAt(0) - 97;
            const b = eq.charCodeAt(3) - 97;
            union(a, b);
        }
    }

    // Then process all "!=" equations
    for (const eq of equations) {
        if (eq[1] === "!") {
            const a = eq.charCodeAt(0) - 97;
            const b = eq.charCodeAt(3) - 97;
            if (find(a) === find(b)) {
                return false;
            }
        }
    }

    return true;
};
